### PR TITLE
Add NLTK: Natural Language Toolkit

### PR DESCRIPTION
**Pull request by msabramo**
[NLTK](https://www.nltk.org/) depends on sqlite3, so this can be merged after https://github.com/iodide-project/pyodide/pull/352

Cc: @madhur-tandon

### CODE REVIEW

**Comment by madhur-tandon** [comment]
@msabramo I am in the process of testing it out. The rebuild of ``cpython`` for ``sqlite3`` takes time. Regardless, I will let you know 👍.

PS - If I might ask, could you also try adding support for ``opencv`` OR ``pillow`` as one of the next steps? I am not much equipped with the knowledge of tinkering with C extensions but I think support for ``opencv`` and ``pillow`` would be great too!

**Comment by madhur-tandon** [comment]
@msabramo Checked it out - works!
NLTK is unable to download models though. @mdboom Will that be supported or is it something that can't be done for now?

**Comment by msabramo** [comment]
@madhur-tandon  Awesome! Thank you for testing! And yes I had the same problem. I struggled to find NLTK functions that didn’t try to download corpora, because all of those failed. I’m not sure how to handle that in a browser environment where things are sandboxed and filesystem access is severely limited. Hoping @mdboom or others might have ideas on how to deal with that, so NLTK can be much more useful. 

Funny you should mention OpenCV and Pillow (Pillow especially). My interests lie in image processing and machine learning applied to images so both of those are on my mind. And Pillow, I got that working locally last night (I had to manually download and build a ton of image libraries; haven’t submitted the PR for that yet; will take a bit of time to get all the image library downloading and building into a `Makefile`). 

So keep an eye out for a Pillow PR in a bit... It will be great to have you test that like you did here.

**Comment by madhur-tandon** [comment]
@msabramo That sounds great! I particularly asked for ``Pillow`` as I have some interesting ideas for different kinds of Demo Iodide Notebooks spanning across different domains, even from classical computer vision such as - K Means Clustering based Segmentation, Harris Corner Detector, etc. [all implemented purely through ``Numpy`` and a bit of ``Matplotlib`` for the results] and the support for ``Pillow`` could enable us easy access to the underlying pixel data of the images (which need to be loaded as ``blobs`` first on the JS side). 

PS - My Interests too lie in the Machine Learning / Data Science Space (+ Software Development), but not limited to Image Processing, a part of me is also inclined towards the aspect of adversarial attacks and robustness. Also, I would love to know more about you and your work. Are you on the ``gitter`` channel for the project? GitHub probably won't be the right medium to talk about all this 😅

**Comment by mdboom** [comment]
> @madhur-tandon Awesome! Thank you for testing! And yes I had the same problem. I struggled to find NLTK functions that didn’t try to download corpora, because all of those failed. I’m not sure how to handle that in a browser environment where things are sandboxed and filesystem access is severely limited. Hoping @mdboom or others might have ideas on how to deal with that, so NLTK can be much more useful.

Do you have examples I can try or tracebacks?  If it's using Python networking APIs, I wouldn't expect it to work due to sandboxing.  In some limited cases, it might be possible to reimplement the parts of the high-level `urlopen` interface on top of Web APIs though -- it would be interesting to see if that would work for NLTK.  Barring that, is there a way to predownload the corpora and ship them as a Pyodide package?

**Comment by msabramo** [comment]
Rebased this to pull in `sqlite` (#352) so that we can try out some real NLTK functions and see how they fail.

**Comment by msabramo** [comment]
Here's an example of a session where I try to use the NLTK word tokenizer and it fails because it wants to download data.

```
Welcome to the Pyodide terminal emulator 🐍
>>> from nltk.tokenize import word_tokenize
>>> text = """This is a test. I'm trying to get nltk working. I hope this works."""
>>> word_tokenize(text)
Error: Traceback (most recent call last):
  File "/lib/python3.7/site-packages/pyodide.py", line 44, in eval_code
    return eval(compile(expr, '<eval>', mode='eval'), ns, ns)
  File "<eval>", line 1, in <module>
  File "/lib/python3.7/site-packages/nltk/tokenize/__init__.py", line 143, in word_tokenize
    sentences = [text] if preserve_line else sent_tokenize(text, language)
  File "/lib/python3.7/site-packages/nltk/tokenize/__init__.py", line 104, in sent_tokenize
    tokenizer = load('tokenizers/punkt/{0}.pickle'.format(language))
  File "/lib/python3.7/site-packages/nltk/data.py", line 868, in load
    opened_resource = _open(resource_url)
  File "/lib/python3.7/site-packages/nltk/data.py", line 993, in _open
    return find(path_, path + ['']).open()
  File "/lib/python3.7/site-packages/nltk/data.py", line 699, in find
    raise LookupError(resource_not_found)
LookupError: 
**********************************************************************
  Resource [93mpunkt[0m not found.
  Please use the NLTK Downloader to obtain the resource:
  [31m>>> import nltk
  >>> nltk.download('punkt')
  [0m
  Attempted to load [93mtokenizers/punkt/PY3/english.pickle[0m
  Searched in:
    - '/home/web_user/nltk_data'
    - '/nltk_data'
    - '/share/nltk_data'
    - '/lib/nltk_data'
    - '/usr/share/nltk_data'
    - '/usr/local/share/nltk_data'
    - '/usr/lib/nltk_data'
    - '/usr/local/lib/nltk_data'
    - ''
**********************************************************************
```

**Comment by msabramo** [comment]
This, on the other hand, works:

```
>>> import nltk
>>> print(nltk.FreqDist(['foo', 'foo', 'foo', 'foo', 'bar', 'bar', 'dog']))
<FreqDist with 3 samples and 7 outcomes>
```

**Comment by msabramo** [comment]
These work too:

```pycon
>>> print(nltk.edit_distance('foo', 'food'))
1
```

```pycon
>>> print(nltk.jaccard_distance(set('mapping'), set('mappings')))
0.14285714285714285
```

**Comment by mdboom** [comment]
It seems like you could probably use [these instructions](https://www.nltk.org/data.html) to download all of the data server-side as part of the build.  Then insert those into package's filesystem contents at a place that nltk could find them.  The `meta.yaml` format supports individual files in the `source/extras` section (see matplotlib), but it could be extended to handle entire directories, which you'd probably want to do in this case.

File: test/test_nltk.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    assert edit_distance == 1
+
+
+def test_nltk_jaccard_distance(selenium):
+    selenium.load_package('nltk')
+    selenium.run('import nltk')
+    jaccard_distance = selenium.run("""
+        nltk.jaccard_distance(set('mapping'), set('mappings'))
+    """)
+    assert jaccard_distance - 0.1428571 < 0.0001
```
Comment (rth) [inline test/test_nltk.py]: Better to add an `abs`  (or use `pytest.approx`).

**Comment by rth** [comment]
(I think handling data downloads could be done at some later point)

**Comment by matham** [comment]
I'm trying to port a library that depends on `pillow` and I'm wondering if you made any progress there or if you ever pushed a branch that I can test with?

File: test/test_nltk.py (hunk 1 above)
Comment (msabramo) [reply test/test_nltk.py]: Thanks! I went ahead and changed it to use [`pytest.approx`](https://docs.pytest.org/en/latest/reference.html#pytest-approx) (which was new to me, so thanks for pointing it out!).

**Comment by msabramo** [comment]
@rth, @mdboom: Is this something you feel comfortable merging? Or would you like me to make some more changes?

**Comment by msabramo** [comment]
@rth: Did the updates you requested. 🤞 that the CI builds all are green.

**Comment by rth** [comment]
Thanks @msabramo !

**Pull request merged**